include/LightGBM/c_api.h

```diff
@@ -1321,7 +1328,9 @@ LIGHTGBM_C_EXPORT int LGBM_NetworkInitWithFunctions(int num_machines,
                                                     void* reduce_scatter_ext_fun,
                                                     void* allgather_ext_fun);
 
-#if defined(_MSC_VER)
+#ifndef __cplusplus
+#define THREAD_LOCAL  /*!< \brief Thread local specifier no-op in C builds. */
```

Comment:
It seems C11 also has thread_local, refer to https://en.wikipedia.org/wiki/C11_(C_standard_revision)

Reply:
The follow-up changes take advantage of `_Thread_local` from C11. They also make the C API header compatible with ANSI C. If the changes to the comments disrupt doc generation, they could be rolled back in favor of abandoning support for C versions prior to C99. The attached [test.c.gz](https://github.com/microsoft/LightGBM/files/7229809/test.c.gz) can be gunzipped in `c_api_test/` and used to test both code generation and compilation with the below:

Code generation:
`gcc test.c -ansi -I../../include/ -E` - should exclude `inline` and `_Thread_local`
`gcc test.c -std=c99 -I../../include/ -E` - should include `inline` and exclude `_Thread_local`
`gcc test.c -std=c11 -I../../include/ -E` - should include `inline` and `_Thread_local`
`g++ test.c -I../../include/ -E` - should include `inline` and `thread_local`

Remove the `-E` switch on the above to confirm compilation for each setting.